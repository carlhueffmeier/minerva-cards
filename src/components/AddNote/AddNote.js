import React from 'react';
import PropTypes from 'prop-types';
import { DeckSelect, TypeSelect, FieldInput, BackButton } from 'components';

AddNote.propTypes = {
  onSave: PropTypes.func.isRequired,
  deckId: PropTypes.string.isRequired,
  onDeckChange: PropTypes.func.isRequired,
  typeId: PropTypes.string.isRequired,
  onTypeChange: PropTypes.func.isRequired,
  fieldValues: PropTypes.object.isRequired,
  onFieldChange: PropTypes.func.isRequired
};

function AddNote(props) {
  function handleSubmit(event) {
    event.preventDefault(); // Prevent page refresh
    props.onSave();
  }

  return (
    <div>
      <BackButton />
      <form onSubmit={handleSubmit}>
        <DeckSelect value={props.deckId} onChange={props.onDeckChange} />
        <TypeSelect value={props.typeId} onChange={props.onTypeChange} />
        <FieldInput
          typeId={props.typeId}
          values={props.fieldValues}
          onFieldChange={props.onFieldChange}
        />
        <button type="submit">Save</button>
      </form>
    </div>
  );
}

export default AddNote;
