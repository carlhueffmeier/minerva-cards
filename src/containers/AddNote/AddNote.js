import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DeckSelect } from 'components';

export default class AddNote {
  state = {
    deckId: this.props.match.params.deckId || null
  };
  render() {
    return (
      <div>
        <form>
          <DeckSelect
            value={this.state.deckId}
            onChange={id => this.setState({ deckId: id })}
          />
          <label>
            Question<input />
          </label>
          <label>
            Answer<input />
          </label>
        </form>
      </div>
    );
  }
}
