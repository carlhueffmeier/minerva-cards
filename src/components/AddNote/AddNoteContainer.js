import React, { Component } from 'react';
import { connect } from 'react-redux';
import { operations as noteOperations } from 'state/modules/notes';
import { selectors as typeSelectors } from 'state/modules/types';
import AddNote from './AddNote';

class AddNoteContainer extends Component {
  state = {
    deckId: this.props.match.params.deckId || ``,
    typeId: this.props.initialType,
    fields: {}
  };

  handleSave() {
    const newNote = {
      deck: this.state.deckId,
      type: this.state.typeId,
      fields: this.state.fields
    };
    this.props.addNote(newNote); // Save to redux store
    this.setState({ fields: {} });
  }

  handleFieldChange(field, value) {
    this.setState(state => ({
      ...state,
      fields: {
        ...state.fields,
        [field]: value
      }
    }));
  }

  handleDeckChange(deckId) {
    this.setState({ deckId });
  }

  handleTypeChange(typeId) {
    this.setState({ typeId });
    this.setState({ fields: {} });
  }

  render() {
    return (
      <AddNote
        onSave={this.handleSave.bind(this)}
        deckId={this.state.deckId}
        onDeckChange={this.handleDeckChange.bind(this)}
        typeId={this.state.typeId}
        onTypeChange={this.handleTypeChange.bind(this)}
        fieldValues={this.state.fields}
        onFieldChange={this.handleFieldChange.bind(this)}
      />
    );
  }
}

function mapStateToProps(state) {
  return {
    // This SHOULD be the default type for this deck (future improvement)
    initialType: typeSelectors.getAllTypes(state)[0].id
  };
}

export default connect(mapStateToProps, { addNote: noteOperations.addNote })(
  AddNoteContainer
);
