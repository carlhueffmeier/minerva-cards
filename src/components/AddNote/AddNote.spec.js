import React from 'react';
import { shallow } from 'enzyme';
import AddNote from './AddNote';

describe('AddNote', () => {
  const props = {
    onSave: () => null,
    deckId: `deck-1`,
    onDeckChange: () => null,
    typeId: `type-1`,
    onTypeChange: () => null,
    fieldValues: {},
    onFieldChange: () => null
  };
  const addNote = shallow(<AddNote {...props} />);

  it('renders properly', () => {
    expect(addNote).toMatchSnapshot();
  });
});
