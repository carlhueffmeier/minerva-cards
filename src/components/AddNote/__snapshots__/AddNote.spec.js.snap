// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddNote
    deckId="deck-1"
    fieldValues={Object {}}
    onDeckChange={[Function]}
    onFieldChange={[Function]}
    onSave={[Function]}
    onTypeChange={[Function]}
    typeId="type-1"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <withRouter(BackButtonContainer) />,
        <form
          onSubmit={[Function]}
>
          <Connect(DeckSelectContainer)
                    onChange={[Function]}
                    value="deck-1"
          />
          <Connect(TypeSelectContainer)
                    onChange={[Function]}
                    value="type-1"
          />
          <Connect(FieldInputContainer)
                    onFieldChange={[Function]}
                    typeId="type-1"
                    values={Object {}}
          />
          <button
                    type="submit"
          >
                    Save
          </button>
</form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Connect(DeckSelectContainer)
              onChange={[Function]}
              value="deck-1"
/>,
            <Connect(TypeSelectContainer)
              onChange={[Function]}
              value="type-1"
/>,
            <Connect(FieldInputContainer)
              onFieldChange={[Function]}
              typeId="type-1"
              values={Object {}}
/>,
            <button
              type="submit"
>
              Save
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onChange": [Function],
              "value": "deck-1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onChange": [Function],
              "value": "type-1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onFieldChange": [Function],
              "typeId": "type-1",
              "values": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Save",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Save",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <withRouter(BackButtonContainer) />,
          <form
            onSubmit={[Function]}
>
            <Connect(DeckSelectContainer)
                        onChange={[Function]}
                        value="deck-1"
            />
            <Connect(TypeSelectContainer)
                        onChange={[Function]}
                        value="type-1"
            />
            <Connect(FieldInputContainer)
                        onFieldChange={[Function]}
                        typeId="type-1"
                        values={Object {}}
            />
            <button
                        type="submit"
            >
                        Save
            </button>
</form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Connect(DeckSelectContainer)
                onChange={[Function]}
                value="deck-1"
/>,
              <Connect(TypeSelectContainer)
                onChange={[Function]}
                value="type-1"
/>,
              <Connect(FieldInputContainer)
                onFieldChange={[Function]}
                typeId="type-1"
                values={Object {}}
/>,
              <button
                type="submit"
>
                Save
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onChange": [Function],
                "value": "deck-1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onChange": [Function],
                "value": "type-1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onFieldChange": [Function],
                "typeId": "type-1",
                "values": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Save",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Save",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
